modules/kafka/tests/test_kafka_topic.c

```diff
@@ -22,157 +22,181 @@
+
+  g_assert_null(g_hash_table_lookup(kafka_driver->topics, "validtopic"));
 
-  cr_assert_str_eq(kafka_driver->topic_name->template, "$MSG");
-  worker = (KafkaDestWorker *) kafka_dest_worker_new(&kafka_driver->super, 0);
   LogMessage *msg = log_msg_new_empty();
+  log_msg_set_value_by_name(msg, "MSG", "validtopic", -1);
+  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic_from_template(worker, msg)), "validtopic");
+  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic_from_template(worker, msg)),
+                   rd_kafka_topic_name(_calculate_topic(worker, msg)));
 
-  log_msg_set_value_by_name(msg, "MSG", "valid_template_topic", -1);
-  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic(worker, msg)), "valid_template_topic");
-  log_msg_set_value_by_name(msg, "MSG", "valid.template.topic", -1);
-  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic(worker, msg)), "valid.template.topic");
-  log_msg_set_value_by_name(msg, "MSG", "valid-template-topic", -1);
-  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic(worker, msg)), "valid-template-topic");
+  g_assert_nonnull(g_hash_table_lookup(kafka_driver->topics, "validtopic"));
+
+  log_msg_set_value_by_name(msg, "MSG", "invalid name", -1);
+
+  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic_from_template(worker, msg)), "fallbackhere");
+  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic_from_template(worker, msg)),
+                   rd_kafka_topic_name(_calculate_topic(worker, msg)));
+  log_msg_unref(msg);
 
   log_threaded_dest_worker_free(&worker->super);
+  log_pipe_deinit(&driver->super);
+  log_pipe_unref(&driver->super);
+  cfg_free(configuration);
 }
 
-Test(kafka_topic, test_invalid_template_calculate_topic)
+Test(kafka_topic, test_get_literal_topic)
 {
-  _init_valid_template_topic_names();
-  kafka_driver->kafka = _construct_client(kafka_driver);
-  _init_template_topic_name(kafka_driver);
+  configuration = cfg_new_snippet();
+  LogDriver *driver = kafka_dd_new(configuration);
 
-  cr_assert_str_eq(kafka_driver->topic_name->template, "$MSG");
-  worker = (KafkaDestWorker *) kafka_dest_worker_new(&kafka_driver->super, 0);
-  LogMessage *msg = log_msg_new_empty();
+  LogTemplate *topic_name = log_template_new(configuration, NULL);
+
+  log_template_compile(topic_name, "topicname", NULL);
+  kafka_dd_set_topic(driver, topic_name);
+
+  g_assert(kafka_dd_init((LogPipe *) driver));
+
+  KafkaDestDriver *kafka_driver = (KafkaDestDriver *) driver;
 
-  log_msg_set_value_by_name(msg, "MSG", "invalid topic", -1);
-  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic(worker, msg)), "fallbacktopicname");
-  log_msg_set_value_by_name(msg, "MSG", "..", -1);
-  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic(worker, msg)), "fallbacktopicname");
-  log_msg_set_value_by_name(msg, "MSG", "", -1);
-  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic(worker, msg)), "fallbacktopicname");
+  g_assert_null(kafka_driver->topics);
```

Comment:
This is implementation detail, also you should not use `g_assert()`. Use `cr_assert_not(_topic_name_is_a_template(kafka_driver));`

Reply:
This is not fixed, don't mark it resolved.